titus-supplementary-component/task-relocation/src/main/java/com/netflix/titus/supplementary/relocation/connector/KubernetesNodeDataResolver.java

```diff
@@ -0,0 +1,115 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netflix.titus.supplementary.relocation.connector;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.function.Function;
+import java.util.function.Predicate;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+import com.netflix.titus.common.util.RegExpExt;
+import com.netflix.titus.runtime.connector.kubernetes.KubeApiFacade;
+import com.netflix.titus.supplementary.relocation.RelocationConfiguration;
+import io.kubernetes.client.informer.SharedIndexInformer;
+import io.kubernetes.client.openapi.models.V1Node;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import static com.netflix.titus.runtime.kubernetes.KubeConstants.NODE_LABEL_MACHINE_GROUP;
+import static com.netflix.titus.runtime.kubernetes.KubeConstants.TAINT_EFFECT_NO_EXECUTE;
+
+public class KubernetesNodeDataResolver implements NodeDataResolver {
+
+    private static final Logger logger = LoggerFactory.getLogger(KubernetesNodeDataResolver.class);
+
+    private static final long NOT_SYNCED_STALENESS_MS = 10 * 3600_000;
+
+    private final SharedIndexInformer<V1Node> nodeInformer;
+    private final Predicate<V1Node> nodeFilter;
+
+    private final Function<String, Matcher> relocationRequiredTaintsMatcher;
+    private final Function<String, Matcher> relocationRequiredImmediatelyTaintsMatcher;
+
+    public KubernetesNodeDataResolver(RelocationConfiguration configuration,
+                                      KubeApiFacade kubeApiFacade,
+                                      Predicate<V1Node> nodeFilter) {
+        this.nodeInformer = kubeApiFacade.getNodeInformer();
+        this.relocationRequiredTaintsMatcher = RegExpExt.dynamicMatcher(
+                configuration::getNodeRelocationRequiredTaints,
+                "nodeRelocationRequiredTaints",
+                Pattern.DOTALL,
+                logger);
+        this.relocationRequiredImmediatelyTaintsMatcher = RegExpExt.dynamicMatcher(
+                configuration::getNodeRelocationRequiredImmediatelyTaints,
+                "nodeRelocationRequiredImmediatelyTaints",
+                Pattern.DOTALL,
+                logger);
+        this.nodeFilter = nodeFilter;
+    }
+
+    @Override
+    public Map<String, Node> resolve() {
+        List<V1Node> k8Nodes = nodeInformer.getIndexer().list().stream().filter(nodeFilter).collect(Collectors.toList());
+        Map<String, Node> result = new HashMap<>();
+        k8Nodes.forEach(k8Node -> toReconcilerNode(k8Node).ifPresent(node -> result.put(node.getId(), node)));
```

Comment:
minor : could it be combined with the indexer -> stream -> filter expression above ?

Reply:
I wanted separate data access and filtering from data transformation. It helps in the debugging process.